simple_salesforce/api.py

```diff
@@ -423,6 +444,115 @@ def _call_salesforce(self, method, url, **kwargs):
+        """Location of the template file"""
+        return os.path.join(self.location, template_file_name)
+
+    def _close_job(self, job_id):
+        """Makes Salesforce call to close a job"""
+        with open(self._template_location('close_job.xml'),
+                  'r') as close_xml_file:
+            close_xml_string = close_xml_file.read()
+            url = self.bulk_url + "/" + job_id
+            response, \
+                response_object = self._make_bulk_request(url,
+                                                          close_xml_string)
+
+            if 'jobInfo' not in response_object or \
+               'state' not in response_object['jobInfo'] or \
+               response_object['jobInfo']['state'] != "Closed":
+                raise SalesforceGeneralError(url,
+                                             "close",
+                                             response.status_code,
+                                             response.text)
+
+    def _create_job(self, data):
+        """Makes a bulk api request to create a job"""
+        url = self.bulk_url
+        job_info = self._make_bulk_request(url, data)
+
+        return job_info[1]['jobInfo']['id']
+
+    def _create_basic_job(self, object_name, method_type):
+        """Create a basic job, eg. insert"""
+        with open(self._template_location('basic_job_xml_template.xml'),
+                  'r') as basic_job_template_file:
+            basic_job_template_string = basic_job_template_file.read()
+            basic_job_xml = basic_job_template_string.format(method_type,
+                                                             object_name)
+            return self._create_job(basic_job_xml)
+
+    def _create_upsert_job(self, object_name, external_field):
+        """Create an upsert job"""
+        with open(self._template_location('upsert_job_xml_template.xml'),
+                  'r') as upsert_job_xml_template_file:
+            upsert_job_template_string = upsert_job_xml_template_file.read()
+            upsert_job_xml = upsert_job_template_string.format(
+                object_name,
+                external_field)
+            return self._create_job(upsert_job_xml)
+
+    def _create_batch(self, csv_file, job_id):
+        """Create a batch from a csv file"""
+        csv_file_data = open(csv_file, 'rb').read()
+        url = self.bulk_url + "/" + job_id + "/batch"
+        self.request.request(
+            "POST",
+            url,
+            headers=self.batch_headers,
+            data=csv_file_data)
+
+    def bulk_insert(self, object_name, csv_file):
+        """Bulk upload a CSV to a Salesforce Object
+
+        Arguments:
+
+        * object_name -- name of the object to upload to
+        * csv_file -- data to upload to the object"""
+
+        job_id = self._create_basic_job(object_name, "insert")
+        self._create_batch(csv_file, job_id)
+        self._close_job(job_id)
```

Comment:
Should jobs *always* be closed? What if there's a socket error or something unexpected that happened in `_create_batch`? Should this perhaps be in a try/except block?